### PR TITLE
Fix: Autofix quotes produces invalid javascript (fixes #4380)

### LINKED ISSUE #4380: Autofix quotes produces invalid javascript
- when using `double` option, `'\\'` is converted to `"\"` instead of `"\\"`
- when using `backtick` option, `${` is not escaped and `'foo${bar}baz'` becomes <code>\`foo${bar}baz\`</code>  instead of <code>\`foo\${bar}baz\`</code>

- update: converting `"don\'t"` to single quote doesn't work as well

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/28144332-autofix-quotes-produces-invalid-javascript?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by nightwing**

### CODE REVIEW

File: lib/rules/quotes.js [hunk 1]
```diff
@@ -23,29 +21,36 @@ var QUOTE_SETTINGS = {
...
         alternateQuote: "\"",
-        description: "backtick",
-        convert: function(str) {
-            return str.replace(/`/g, "\`").replace(/^(?:\\*)["']|(?:\\*)["']$/g, "`");
-        }
+        description: "backtick"
     }
 };
 
+QUOTE_SETTINGS.double.convert =
```
Comment (nzakas) [inline lib/rules/quotes.js]: Can you add some comments to this function explaining what it's doing?

**Comment by IanVS** [comment]
:+1: https://github.com/eslint/eslint/issues/3090 will be easier to address if all three types have the same `convert` function, as you have them.

**Comment by nightwing** [comment]
Do we need to support converting multiline backticked strings to quoted strings, or replacing `\n\` followed by newline with a newline when converting quoted string to backticked one?

```js
`line1
line2`

"line1\n\
line2"
```

**Comment by nzakas** [comment]
We should never create multi-line strings. If you have:

```js
`line1
line2`;
```

It should become

```js
"line1\nline2";
```

**Comment by nightwing** [comment]
Ok, added handling of newlines for that case.

**Comment by nzakas** [comment]
Can you add some tests for the newline case?

**Comment by nightwing** [comment]
Currently no rule triggers that, and if i understand correctly #3090 is not supposed to trigger that either. So to test this i'll have to export QUOTE_SETTINGS and call it directly. Do you think that's an acceptable solution?

**Comment by nzakas** [comment]
I don't understand, why can't you just add the test case I showed above?

**Comment by nightwing** [comment]
because quotes rule checks only for "Literal" nodes https://github.com/eslint/eslint/blob/791ed1a06fd75cbb237c725eaef36404f42d1948/lib/rules/quotes.js#L142 and is never called for "TemplateLiteral" nodes.

Additionally https://github.com/eslint/eslint/issues/3090#issuecomment-141575784 asks to not convert multiline template literals to quoted strings, so there is no way to trigger the above conversion from a testcase.

**Comment by nzakas** [comment]
Okay, we probably need to revisit, but this looks good for now.

**Pull request merged**